### PR TITLE
ogl bringup part 2

### PR DESCRIPTION
**Pull request by kd-11**
Bugfixes and optimizations. Followup to https://github.com/RPCS3/rpcs3/pull/12114
- Implements fast(er) texture readback on AMD OGL using compute shaders for some commonly bitcasted formats
- Fix incorrect size reporting on destroyed buffer object
- Fix texture binding bugs in the overlay
- Refactors buffer objects to make things easier to manage. The whole OGL backend is getting refactored later, this just happened to block some other work I was doing so I chose to do it now.
- Implemented scratch buffers (for ssbo usage) with CPU-side barriers to track collision. This essentially moves a bunch of driver logic to our side, giving us more control and making it easier to profile and optimize some things.
- Fix GPU deswizzling. I don't think it even worked at all with the previous PR, some paths were completely broken.
- Restore some missing depth-format conversions accidentally removed in the last PR

Fixes https://github.com/RPCS3/rpcs3/issues/12148
Fixes https://github.com/RPCS3/rpcs3/issues/12145
Fixes https://github.com/RPCS3/rpcs3/issues/12142
Fixes https://github.com/RPCS3/rpcs3/issues/12143
Fixes https://github.com/RPCS3/rpcs3/issues/12149

### CODE REVIEW

**Comment by Darkhost1999** [comment]
ACR crash fixed, ACR face with (image) all over Ezio's face is fixed. waiting on depth commit for further comments
pic

File: rpcs3/Emu/RSX/GL/glutils/ring_buffer.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		void create(buffer::target _target, u64 size);
+		void remove();
+
+		u32 alloc(u32 size, u32 alignment);
+		void push_barrier(u32 start, u32 length);
+
+		buffer& get() { return m_storage; }
+		u64 size() const { return m_storage.size(); }
+	};
+}
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/glutils/ring_buffer.h:113]: missing newline

**Comment by Darkhost1999** [comment]
All I'm getting is this. Unless I need a special setting IDK. This took me like 12 times to get and the game froze when the camera was moved and the compass moves that tiny bit while audio keeps going.
I'll probably need to open a ticket after this PR merges. IDK

https://user-images.githubusercontent.com/60384196/171758008-05c7d4a0-8dd8-46d3-8fe1-e340df0580fc.mp4

The game was only booted once by me in this entire log
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/8828435/RPCS3.log)

**Comment by Darkhost1999** [comment]
Everytime I use a new PR I get a fatal error blame windows. Clearing all caches fixes the issue. So I'm using brand new caches when testing.

**Comment by Realmantik** [comment]
RE Code Veronica still not fixed.
[NPUB30467_20220603105149_capture.zip](https://github.com/RPCS3/rpcs3/files/8830587/NPUB30467_20220603105149_capture.zip)
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/8830657/RPCS3.log.gz)

**Comment by kd-11** [comment]
> All I'm getting is this. Unless I need a special setting IDK. This took me like 12 times to get and the game froze when the camera was moved and the compass moves that tiny bit while audio keeps going. I'll probably need to open a ticket after this PR merges. IDK
> ACR.mp4
> 
> The game was only booted once by me in this entire log [RPCS3.log](https://github.com/RPCS3/rpcs3/files/8828435/RPCS3.log)

Yes, open a ticket for that.

**Comment by kd-11** [comment]
> RE Code Veronica still not fixed. [NPUB30467_20220603105149_capture.zip](https://github.com/RPCS3/rpcs3/files/8830587/NPUB30467_20220603105149_capture.zip) [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/8830657/RPCS3.log.gz)

I'll also need a ticket for this. Since master is broken for many games, I'll have to prioritize getting this in quickly to avoid affecting too many users negatively.

**Comment by z0z0z** [comment]
Thanks kd-11 this PR fixes the crashing with
```
E RSX: ERROR: 0:44: 'f32_to_d24' : no matching overloaded function found 
ERROR: 0:44: 'assign' : cannot convert from ' const float' to ' temp highp uint'
```
Was probably commit d2d2b4dc7fed64cc858567c4c9c204908f54ac68

**Pull request merged**